### ISSUE TITLE
XaAES - AES API conformance

### ISSUE THREAD

**Issue by xdelatour** [freemint/freemint#299]
### Missing features

AES
- [ ] appl_read
- [ ] appl_tplay/appl_trecord
- [ ] graf_movebox/graf_growbox/graf_shrinkbox
- [ ] rsrc_saddr
- [ ] shel_put/shel_get

MagiC
- [ ] form_xerr

MagiC/PC-GEM
- [ ] appl_bvset
- [ ] menu_unregister
- [ ] menu_click
- [ ] scrp_clear

N.AES
- [ ] graf_multirubber
- [ ] shel_help
- [ ] wind_draw

N.AES/PC-GEM
- [ ] shel_rdef/shel_wdef

FreeGEM
- [ ] appl_init(X_BUF_V2)/appl_xbvset/appl_xbvget
- [ ] proc_create/proc_run/proc_delete/proc_info/proc_malloc/proc_free/proc_switch/proc_shrink/proc_setblock
- [ ] prop_get/prop_put/prop_del/prop_gui_get/prop_gui_set
- [ ] xgrf_stepcalc/xgrf_2box/xgrf_color/xgrf_dtimage
- [ ] xshl_getshell/xshl_setshell

MyAES
- [ ] appl_getcicon

MultiGEM
- [ ] wind_apfind

Geneva
- [ ] All functions

#### Missing wind_get subfunctions

- [ ] (MagiC) WF_M_WINDLIST/WF_INFOXYWH
- [ ] (N.AES) WF_WIDGETS
- [ ] (MyAES) WF_USER_POINTER/WF_WIND_ATTACH/WF_BITMAP
- [ ] (FreeGEM) WF_OBFLAG/WF_OBTYPE/WF_OBSPEC
- [ ] (WINX) WF_WINX/WF_WINXCFG/WF_DDELAY
- [ ] (Geneva) All subfunctions

#### Missing wind_set subfunctions
- [ ] (AES) WF_COLOR/WF_DCOLOR
- [ ] (MagiC) WF_KIND
- [ ] (FreeGEM) WF_OBFLAG/WF_OBSPEC
- [ ] (N.AES) WF_WIDGETS/WF_APPICON
- [ ] (MyAES) WF_USER_POINTER/WF_WIND_ATTACH/WF_OPTIONS/WF_FULLSCREEN
- [ ] (WINX) WF_WINXCFG/WF_DDELAY/WF_STACK/WF_TOPALL/WF_BOTTOMALL
- [ ] (Geneva) All subfunctions

###  Incomplete/bugged/unavailable features
- [ ] toolbar support #230 #231
- [ ] popup menus/scrolling menus/hierarchical menus/dropdown lists menus #243 #262
- [ ] No scroll list API
- [ ] wind_create with [xkind (32bit)](https://github.com/freemint/freemint/blob/b505c3c7e0b30afdac83fa7d5a857bb4f91efad5/xaaes/src.km/newcalls.txt#L424-L439)
- [ ] application icon widget in titlebar
- [ ] context menu widget in titlebar
- [ ] shel_write 20/21/22 (threads)
- [ ] color palette in RSC files
- [ ] New event/poll functions
- [ ] TORG compliance (101-109)

**Comment by mikrosk** [comment]
Great list, Xavier. Many thanks for that!

**Comment by th-otto** [comment]
Yes, great list. But imho, some things can immediately be solved:

- it doesn't make sense to do anything about things that are only supported by FreeGEM. No atari program will ever use it. That applies to the wind_get() subfunctions, the proc_* calls etc. Those things are only documented in tos.hyp for completeness, and maybe also to avoid using the same opcodes for anything else.
- Supporting WF_WINX and/or WF_WINXCFG doesn't make much sense either, since XaAES ist not WINX.

**Comment by xdelatour** [comment]
> the fslx_xxx functions are implemented, just disabled by default

Looks like fslx_xxx functions are [dummy functions](https://github.com/freemint/freemint/blob/master/xaaes/src.km/xa_fslx.c)
 
Perhaps adding other functions may improve compatibility with some programs (I'm pretty sure I've seen programs use appl_read to read extra bytes one by one, but I can't remember where)

**Comment by xdelatour** [comment]
Patch to add form_xerr() (changes to existing form_error()):

[form_xerr-20230418-v3.patch.txt](https://github.com/freemint/freemint/files/11260810/form_xerr-20230418-v3.patch.txt)

Checked with MagiC PC, contrary to form_error(), form_xerr() returns the errcode instead of 1 (button number).

In the above post, patch to add wing_get(WF_INFOXYWH)

**Comment by th-otto** [comment]
I don't think we should support WF_M_WINDLIST. It is really very MagiC specific, and the only application that i know of that makes use of it is mgnotice, and even there the code is commented out. Typically, applications maintain their own list of managed windows, anyway.

**Comment by mikrosk** [comment]
@xdelatour So what about this patch? Obsolete? (as mentioned in https://github.com/freemint/freemint/issues/293#issuecomment-1520666261, not sure what is the outcome :))

**Comment by Landemarre** [comment]
Hello

I'm currently adding a small feature in MyAES that would be nice to have too in XaAES, I'm going to use it in Thing 1.60, this is only cosmetic, I have mix 2 object state to add a new symbol link to object like CHECKED. I have mix CHECKED with CROSSED to display a red cross in place of check and cross object, see photo after.
![checked_crossed](https://github.com/freemint/freemint/assets/25240527/297e08a7-8906-412c-b1a5-4f7fa4e51ba3)

Notice I think check symbol and cross to my point of view with icon can be not at nice position (check always on top left and cross at size of object, my point of view better display check should be at left top of icon in the object of course when possible and cross should cross only the icon.
